man/trisetcmp.1
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "TRISETCMP" "1" "10 June 2026" "" "The Regina Handbook"

.SH NAME
trisetcmp \- Compare triangulations between two Regina data files
.SH SYNOPSIS

\fBtrisetcmp\fR [ \fB-m\fR | \fB-n\fR ] [ \fB-s\fR ] \fB\fIfile1\fB\fR \fB\fIfile2\fB\fR


\fBtrisetcmp\fR { \fB-v, --version\fR | \fB-?, --help\fR }

.SH "DESCRIPTION"
.PP
Compares all triangulations in the first file against
all triangulations in the second file, looking for pairs of
triangulations that are combinatorially isomorphic.
.PP
The two given files must be Regina data files.
A full list of matches (or a full list of non-matches if
\fB-n\fR is passed) is written to standard output.
A match occurs when some triangulation from
\fIfile1\fR is combinatorially isomorphic to
some triangulation from \fIfile2\fR (i.e.,
identical up to a relabelling of tetrahedra and their vertices).
.PP
This utility can also do subcomplex testing instead of full
isomorphism testing.  See the option \fB-s\fR for
details.
.PP
This utility now tests triangulations of all dimensions (as opposed to
Regina\~<=\~7.4.1, where it only compared triangulations
in dimensions 3 or 4).
However, it will only ever try to match pairs of triangulations of the
\fBsame\fR dimension - for example, it will not test
whether a 3-manifold triangulation from \fIfile1\fR
appears as a subcomplex of a 4-manifold triangulation from
\fIfile2\fR\&.
.PP
In dimension\~3, this utility only looks at Regina's own native
triangulations, and ignores any hybrid  triangulations.
.SH "OPTIONS"
.TP
\fB-m (default)\fR
Output matches only.  All isomorphic matches between
triangulations in \fIfile1\fR
and triangulations in \fIfile2\fR
will be listed.
.TP
\fB-n\fR
Output non-matches only.  All triangulations from
\fIfile1\fR with no isomorphic match in
\fIfile2\fR will be listed, and vice versa.

If \fB-s\fR is passed then non-matches are
tested in one direction only, not both; see below for details.
.TP
\fB-s\fR
Instead of testing triangulations for isomorphism, test whether
one triangulation is isomorphic to a subcomplex of the other.

In the default case of \fB-m\fR (output matches only),
this program outputs all instances where a triangulation from
\fIfile1\fR is isomorphic to a subcomplex
of a triangulation from \fIfile2\fR\&.

In the case of \fB-n\fR (output non-matches only),
this program outputs all triangulations from
\fIfile1\fR that are not isomorphic to a
subcomplex of any triangulation from \fIfile2\fR\&.
.TP
\fB-v, --version\fR
Show which version of Regina is being used, and exit
immediately.
.TP
\fB-?, --help\fR
Display brief usage information, and exit immediately.
.SH "INTERNATIONALISATION"
.PP
If any packets contain international characters,
Regina will attempt to convert these to your local character
encoding as it writes them to the output.
.PP
You can tell Regina what character encoding to use by setting
standard locale-related environment variables, such as
\fBLANG\fR, \fBLC_CTYPE\fR or
\fBLC_ALL\fR\&.
.PP
For example, if \fBLANG\fR is set to
en_AU then output will be written in the
Western European character set ISO-8859-1,
and if \fBLANG\fR is set to
en_AU.UTF-8 then output will be written in the
universal character set UTF-8\&.
.PP
Typically these environment variables will already be set for you when
you install your \fBGNU/Linux\fR system, and Regina will just use the
right character set out of the box.  See your \fBGNU/Linux\fR system reference
for further information on supporting different locales.
.SH "MACOS USERS"
.PP
If you downloaded a drag-and-drop app bundle, this utility is
shipped inside it.  If you dragged Regina to the main
Applications folder, you can run it as
/Applications/Regina.app/Contents/MacOS/trisetcmp\&.
.SH "WINDOWS USERS"
.PP
The command-line utilities are installed beneath the
\fIProgram\~Files\fR directory; on some
machines this directory is called
\fIProgram\~Files\~(x86)\fR\&.
You can start this utility by running
c:\\Program\~Files\\Regina\\Regina\~8.0\\bin\\trisetcmp.exe\&.
.SH "SEE ALSO"
.PP
censuslookup,
regina-gui\&.
.SH "AUTHOR"
.PP
This utility was written by Benjamin Burton
<bab@maths.uq.edu.au>\&.
Many people have been involved in the development
of Regina; see the users' handbook for a full list of credits.


utils/trisetcmp.cpp

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  Compare two data files in search of isomorphic triangulations         *
 *                                                                        *
 *  Copyright (c) 1999-2026, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  As an exception, when this program is distributed through (i) the     *
 *  App Store by Apple Inc.; (ii) the Mac App Store by Apple Inc.; or     *
 *  (iii) Google Play by Google Inc., then that store may impose any      *
 *  digital rights management, device limits and/or redistribution        *
 *  restrictions that are required by its terms of service.               *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public License     *
 *  along with this program. If not, see <https://www.gnu.org/licenses/>. *
 *                                                                        *
 **************************************************************************/

#include <cstdlib>
#include <cstring>
#include "triangulation/dim2.h"
#include "triangulation/dim3.h"
#include "triangulation/dim4.h"
#include "triangulation/hidim.h"
#include "utilities/exception.h"
#include "utilities/i18nutils.h"

using regina::Packet;
using regina::Triangulation;

template <int dim> requires (regina::supportedDim(dim))
using TriPacket = regina::PacketOf<Triangulation<dim>>;

bool subcomplexTesting = false;

void usage(const char* progName, const std::string& error = std::string()) {
    if (! error.empty())
        std::cerr << error << "\n\n";

    std::cerr << "Usage:\n";
    std::cerr << "    " << progName << " [ -m | -n ] [ -s ] "
        "<file1.rga> <file2.rga>\n"
        "    " << progName << " [ -v, --version | -?, --help ]\n";
    std::cerr << std::endl;
    std::cerr << "    -m : List matches, i.e., triangulations "
        "contained in both files (default)\n";
    std::cerr << "    -n : List non-matches, i.e., triangulations "
        "in one file but not the other\n";
    std::cerr << "    -s : Allow triangulations from file1.rga to be "
        "subcomplexes of\n"
        "         triangulations from file2.rga\n";
    std::cerr << std::endl;
    std::cerr << "    -v, --version : Show which version of Regina "
        "is being used\n";
    std::cerr << "    -?, --help    : Display this help\n";
    exit(1);
}

template <int dim> requires (regina::supportedDim(dim))
bool hasMatchFrom(const TriPacket<dim>& src, const Packet& tree) {
    for (const Packet& p : tree) {
        if (p.type() != src.type())
            continue;

        if (subcomplexTesting) {
            if (src.isContainedIn(static_cast<const TriPacket<dim>&>(p)))
                return true;
        } else {
            if (src.isIsomorphicTo(static_cast<const TriPacket<dim>&>(p)))
                return true;
        }
    }
    return false;
}

template <int dim> requires (regina::supportedDim(dim))
size_t writeMatchesFrom(const TriPacket<dim>& src, const Packet& tree,
        std::ostream& out) {
    size_t matches = 0;
    for (const Packet& p : tree) {
        if (p.type() != src.type())
            continue;

        if (subcomplexTesting) {
            if (src.isContainedIn(static_cast<const TriPacket<dim>&>(p))) {
                out << "    " << src.humanLabel() << "  <=  "
                    << p.humanLabel() << std::endl;
                ++matches;
            }
        } else {
            if (src.isIsomorphicTo(static_cast<const TriPacket<dim>&>(p))) {
                out << "    " << src.humanLabel() << "  ==  "
                    << p.humanLabel() << std::endl;
                ++matches;
            }
        }
    }
    return matches;
}

void runMatches(const Packet& tree1, const Packet& tree2, std::ostream& out) {
    if (subcomplexTesting)
        out << "Matching (isomorphic subcomplex) triangulations:\n"
            << std::endl;
    else
        out << "Matching (isomorphic) triangulations:\n" << std::endl;

    long matches = 0;

    for (const Packet& src : tree1)
        if (int srcDim = regina::isTriangulation(src.type()))
            regina::select_constexpr<2, regina::maxDim() + 1, void>(srcDim,
                [&](auto dim) {
                    matches += writeMatchesFrom(
                        static_cast<const TriPacket<dim>&>(src), tree2, out);
                });

    if (matches == 0)
        out << "No matches found." << std::endl;
    else if (matches == 1)
        out << std::endl << "1 match." << std::endl;
    else
        out << std::endl << matches << " matches." << std::endl;
}

void runNonMatches(const std::string& file1, const Packet& tree1,
        const std::string& file2, const Packet& tree2,
        std::ostream& out) {
    out << "Triangulations in " << file1 << " but not " << file2
        << ":\n" << std::endl;
    long missing = 0;

    bool matched;
    for (const Packet& src : tree1)
        if (int srcDim = regina::isTriangulation(src.type()))
            regina::select_constexpr<2, regina::maxDim() + 1, void>(srcDim,
                [&](auto dim) {
                    if (! hasMatchFrom(static_cast<const TriPacket<dim>&>(src),
                            tree2)) {
                        out << "    " << src.humanLabel() << std::endl;
                        ++missing;
                    }
                });

    if (missing == 0)
        out << "All triangulations matched." << std::endl;
    else if (missing == 1)
        out << std::endl << "1 non-match." << std::endl;
    else
        out << std::endl << missing << " non-matches." << std::endl;
}

int main(int argc, char* argv[]) {
    std::string file1, file2;
    bool listMatches = false;
    bool listNonMatches = false;

    // Parse command line.
    bool noMoreOpts = false;
    char optChar;
    for (int i = 1; i < argc; i++) {
        if ((! noMoreOpts) && *argv[i] == '-') {
            // Standard arguments:
            if (strcmp(argv[i], "-?") == 0 || strcmp(argv[i], "--help") == 0)
                usage(argv[0]);
            if (strcmp(argv[i], "-v") == 0 ||
                    strcmp(argv[i], "--version") == 0) {
                if (argc != 2)
                    usage(argv[0], "Option --version cannot be "
                        "used with any other arguments.");
                std::cout << PACKAGE_BUILD_STRING << std::endl;
                exit(0);
            }

            // Option.
            if (! argv[i][1])
                usage(argv[0], std::string("Invalid option: ") + argv[i]);
            else if (argv[i][2])
                usage(argv[0], std::string("Invalid option: ") + argv[i]);

            // Argument has length 2.
            optChar = argv[i][1];
            if (optChar == 'm')
                listMatches = true;
            else if (optChar == 'n')
                listNonMatches = true;
            else if (optChar == 's')
                subcomplexTesting = true;
            else if (optChar == '-')
                noMoreOpts = true;
            else
                usage(argv[0], std::string("Invalid option: ") + argv[i]);
        } else if (*argv[i]) {
            noMoreOpts = true;

            if (file1.empty())
                file1 = argv[i];
            else if (file2.empty())
                file2 = argv[i];
            else
                usage(argv[0], "You may not pass more than two data files.");
        } else
            usage(argv[0], "Empty arguments are not allowed.");
    }

    // Sanity checking.
    if (listMatches && listNonMatches) {
        usage(argv[0], "Options -n and -m may not be used together.");
    } else if (! (listMatches || listNonMatches)) {
        // List matches by default.
        listMatches = true;
    }

    if (file1.empty() || file2.empty())
        usage(argv[0], "Two data files must be specified.");

    // Open the two data files.
    std::shared_ptr<Packet> tree1, tree2;
    try {
        tree1 = regina::open(file1.c_str());
    } catch (const regina::FileError&) {
        std::cerr << "File " << file1 << " could not be read.\n";
        std::cerr << "Please check that you have permissions to read it.\n";
        return 1;
    } catch (const regina::InvalidInput&) {
        std::cerr << "File " << file1 << " could not be opened.\n";
        std::cerr << "It does not appear to be a Regina data file.\n";
        return 1;
    }
    try {
        tree2 = regina::open(file2.c_str());
    } catch (const regina::FileError&) {
        std::cerr << "File " << file2 << " could not be read.\n";
        std::cerr << "Please check that you have permissions to read it.\n";
        return 1;
    } catch (const regina::InvalidInput&) {
        std::cerr << "File " << file2 << " could not be opened.\n";
        std::cerr << "It does not appear to be a Regina data file.\n";
        return 1;
    }

    // Since we will be writing packet labels on stdout, make sure we use a
    // character encoding that the user can read.
    regina::i18n::IConvStream out(std::cout,
        "UTF-8", regina::i18n::Locale::codeset());

    // Run our tests.
    if (listMatches)
        runMatches(*tree1, *tree2, out);
    else {
        runNonMatches(file1, *tree1, file2, *tree2, out);
        if (! subcomplexTesting) {
            out << std::endl << "--------------------\n" << std::endl;
            runNonMatches(file2, *tree2, file1, *tree1, out);
        }
    }

    // All done.
    return 0;
}

